### ISSUE TITLE
/api/supportMatrix missing some supportedServers and has duplicate values

### ISSUE THREAD

**Issue by rzgry** [eclipse/microprofile-starter#294]
MP32 supported servers returns `KUMULUZEE` 3 times and is missing WildFly and Helidon when compared to https://start.microprofile.io/  web UI

```
curl https://start.microprofile.io/api/supportMatrix
{"configs":{"MP32":{"supportedServers":["LIBERTY","KUMULUZEE","QUARKUS","KUMULUZEE","KUMULUZEE","PAYARA_MICRO","THORNTAIL_V2"],"specs":["CONFIG","FAULT_TOLERANCE","JWT_AUTH","METRICS","HEALTH_CHECKS","OPEN_API","OPEN_TRACING","REST_CLIENT"]},"MP30":{"supportedServers":["HELIDON","KUMULUZEE","THORNTAIL_V2","LIBERTY"],"specs":["CONFIG","FAULT_TOLERANCE","JWT_AUTH","METRICS","HEALTH_CHECKS","OPEN_API","OPEN_TRACING","REST_CLIENT"]},"MP22":{"supportedServers":["THORNTAIL_V2","LIBERTY","HELIDON","PAYARA_MICRO","KUMULUZEE"],"specs":["CONFIG","FAULT_TOLERANCE","JWT_AUTH","METRICS","HEALTH_CHECKS","OPEN_API","OPEN_TRACING","REST_CLIENT"]},"MP21":{"supportedServers":["TOMEE","PAYARA_MICRO","KUMULUZEE","THORNTAIL_V2","LIBERTY"],"specs":["CONFIG","FAULT_TOLERANCE","JWT_AUTH","METRICS","HEALTH_CHECKS","OPEN_API","OPEN_TRACING","REST_CLIENT"]},"MP20":{"supportedServers":["PAYARA_MICRO","TOMEE","KUMULUZEE","LIBERTY"],"specs":["CONFIG","FAULT_TOLERANCE","JWT_AUTH","METRICS","HEALTH_CHECKS","OPEN_API","OPEN_TRACING","REST_CLIENT"]},"MP14":{"supportedServers":["LIBERTY","TOMEE","PAYARA_MICRO","KUMULUZEE"],"specs":["CONFIG","FAULT_TOLERANCE","JWT_AUTH","METRICS","HEALTH_CHECKS","OPEN_API","OPEN_TRACING","REST_CLIENT"]},"MP13":{"supportedServers":["TOMEE","LIBERTY","KUMULUZEE","PAYARA_MICRO","THORNTAIL_V2"],"specs":["CONFIG","FAULT_TOLERANCE","JWT_AUTH","METRICS","HEALTH_CHECKS","OPEN_API","OPEN_TRACING","REST_CLIENT"]},"MP12":{"supportedServers":["PAYARA_MICRO","THORNTAIL_V2","LIBERTY","HELIDON","TOMEE","KUMULUZEE","WILDFLY_SWARM"],"specs":["CONFIG","FAULT_TOLERANCE","JWT_AUTH","METRICS","HEALTH_CHECKS"]}},"descriptions":{"CONFIG":"Configuration - externalize and manage your configuration parameters outside your microservices","OPEN_API":"Open API - Generate OpenAPI-compliant API documentation for your microservices","HEALTH_CHECKS":"Health Checks - Verify the health of your microservices with custom verifications","REST_CLIENT":"Rest Client - Invoke RESTful services in a type-safe manner","FAULT_TOLERANCE":"Fault Tolerance - all about bulkheads, timeouts, circuit breakers, retries, etc. for your microservices","JWT_AUTH":"JWT Propagation - propagate security across your microservices","OPEN_TRACING":"Open Tracing - trace the flow of requests as they traverse your microservices","METRICS":"Metrics - Gather and create operational and business measurements for your microservices"}}
```

```
curl https://start.microprofile.io/api/mpVersion/MP32
{"supportedServers":["THORNTAIL_V2","KUMULUZEE","QUARKUS","KUMULUZEE","PAYARA_MICRO","LIBERTY","KUMULUZEE"],"specs":["CONFIG","FAULT_TOLERANCE","JWT_AUTH","METRICS","HEALTH_CHECKS","OPEN_API","OPEN_TRACING","REST_CLIENT"]}
```

![image](https://user-images.githubusercontent.com/16768710/80530160-9e13f900-8966-11ea-8f2a-b80362c0c581.png)

Occurs for https://start.microprofile.io/api/supportMatrix and same for the versioned endpoints ie  
https://start.microprofile.io/api/3/supportMatrix which is causing duplicates in the MicroProfile Starter VSCode extension. 

![image](https://user-images.githubusercontent.com/16768710/80529730-e252c980-8965-11ea-86b7-62d211f3391f.png)

**Comment by Karm** [comment]
I can confirm this is a problem: 
```
...
        "MP32": {
            "supportedServers": [
                "KUMULUZEE",
                "THORNTAIL_V2",
                "PAYARA_MICRO",
                "QUARKUS",
                "KUMULUZEE",
                "LIBERTY",
                "KUMULUZEE"
            ],
...
```

I will be looking into it.  Right off the bat: The code for the REST API is pretty old and hasn't been changed in a while. On the other hand, WildFly and Kumuluzee servers were recently added/updated and it seems a weird coincidence that these two are also the problematic ones here. I will take a look at a couple of latest commits.

Regarding test suite: Thing like this is not covered...

**Comment by Karm** [comment]
@rdebusscher

 - There must have been something wrong in the [PostConstruct initialization](https://github.com/eclipse/microprofile-starter/blob/v1.2.0-debug/src/main/java/org/eclipse/microprofile/starter/rest/APIService.java#L93:L124), like if it had been called twice (which it shouldn't...) and got into a race? 

 - I wanted to use my weapon of choice when it comes to debugging live production VM and attach [Byteman](https://developer.jboss.org/wiki/ABytemanTutorial#how_do_i_install_the_agent_into_a_running_program), but the container was running on headless JRE without instrumentation lib .so; I switched that to full JDK so as we can debug it in future, i.e. to avoid:
```
libinstrument.so: cannot open shared object file
```

- I run this so as to 94 times restart the app and the error hasn't re-appeared unfortunately  (qjf is a json formatter)

 ```
GOOD=11;while [ 1 ]; do THINGS=`curl  https://start.microprofile.io/api/supportMatrix/ 2>/dev/null | qjf  | grep MP32 -A10 | sort -u | wc -l`;if [ $THINGS -ne $GOOD ]; then echo Broken;break; else echo Restarting;ssh ec2-user@aws-microstarter "sudo systemctl restart docker-compose@start.microprofile.io";fi;sleep 10;done;
```

So, I don't know the root cause and the only explanation I can see is that the PostConstruct was called twice and got into a race.

**Comment by rdebusscher** [comment]
That sound like a bug in Thorntail as ApplicationScoped instances cannot be created/instantiated twice.

**Comment by Karm** [comment]
@rdebusscher Indeed.  The initialization code I use in the init method is obviously not thread safe, but I assumed it doesn't need to be as it was supposed to be called once by a single thread...   I will try to reproduce the issue in a smaller example app....

**Comment by jreznot** [comment]
Also, from https://start.microprofile.io/api/supportMatrix GRAPHQL is not included to any of MP versions
E.g. for MP33:
```
    "MP33": {
      "supportedServers": [
        "LIBERTY",
        "LIBERTY",
        "KUMULUZEE",
        "WILDFLY",
        "HELIDON",
        "WILDFLY"
      ],
      "specs": [
        "CONFIG",
        "FAULT_TOLERANCE",
        "JWT_AUTH",
        "METRICS",
        "HEALTH_CHECKS",
        "OPEN_API",
        "OPEN_TRACING",
        "REST_CLIENT"
      ],
      "specCodes": [
        "CONFIG",
        "FAULT_TOLERANCE",
        "JWT_AUTH",
        "METRICS",
        "HEALTH_CHECKS",
        "OPEN_API",
        "OPEN_TRACING",
        "REST_CLIENT",
        "GRAPHQL"
      ]
    }
``` 

GRAPHQL is in specCodes, but not in specs.

**Comment by Karm** [comment]
@jreznot That would be a new issue I believe.

**Comment by Emily-Jiang** [comment]
@jreznot `specCodes` are the specs that works with MicroProfile versions. `mp GraphQL` is a standalone spec that do not belong to any MP versions. We could use a different heading to differentiate the `specs` and `specCodes`. @Karm @rdebusscher thoughts?

**Comment by Emily-Jiang** [comment]
Based on what @Karm said, the issue reported by @jreznot is different from this issue. I have opened a [new](https://github.com/eclipse/microprofile-starter/issues/403) one to address the concern from @jreznot .

**Comment by Emily-Jiang** [comment]
Close this issue as it is not happening any more.

**Comment by Emily-Jiang** [comment]
<img width="676" alt="image" src="https://user-images.githubusercontent.com/10999780/109638607-1d092d80-7b46-11eb-8023-7e015e8931d5.png">

**Comment by Emily-Jiang** [comment]
Reopen as I hit this issue today.

**Comment by Emily-Jiang** [comment]
I have not seen this lately. Have anyone else?

**Comment by Emily-Jiang** [comment]
@Karm and I discussed this today. We have not seen this for some months. This might be fixed by updating jvm to jdk3 #295 . Close for now. Reopen if it happens again.
